test/spec/features/service/link_spec.rb

```diff
@@ -1,45 +1,75 @@
 describe 'service link' do
-  after(:each) do
-    %w(test-1 test-2).each do |s|
-      run "kontena service rm --force #{s}"
+  context 'with a test-1 service' do
+    before do
+      run! "kontena service create test-1 redis:3.0"
     end
-    run "kontena stack rm --force simple"
-  end
 
-  it 'links service to target' do
-    run "kontena service create test-1 redis:3.0"
-    run "kontena service create test-2 redis:3.0"
-    k = run "kontena service link test-1 test-2"
-    expect(k.code).to eq(0)
-    k = run "kontena service show test-1"
-    expect(k.out.match(/^\s+- test-2\s*$/)).to be_truthy
-  end
+    after do
+      run("kontena service rm --force test-1")
+    end
+
+    context 'with a test-2 service' do
+      before do
+        run! "kontena service create test-2 redis:3.0"
+      end
 
-  it 'links service to stack service' do
-    with_fixture_dir("stack/simple") do
-      run 'kontena stack install --no-deploy'
+      after do
+        run "kontena service unlink test-1 test-2"
+        run "kontena service rm --force test-2"
+      end
+
+      it 'links service to target' do
+        run! "kontena service link test-1 test-2"
+        k = run! "kontena service show test-1"
+        expect(k.out.match(/^\s+- test-2\s*$/)).to be_truthy
+      end
     end
-    run "kontena service create test-1 redis:3.0"
-    k = run "kontena service link test-1 simple/redis"
-    expect(k.code).to eq(0)
-    k = run "kontena service show test-1"
-    expect(k.out.match(/^\s+\- simple\/redis\s*$/)).to be_truthy
-  end
 
-  it 'links stack service with existing links to service' do
-    with_fixture_dir("stack/links") do
-      run 'kontena stack install --no-deploy links.yml'
+    context 'with a simple stack' do
+      before do
+        with_fixture_dir("stack/simple") do
+          run! 'kontena stack install --no-deploy'
+        end
+      end
+
+      after do
+        run! 'kontena stack rm --force simple'
+      end
+      after do
+        run! 'kontena service unlink test-1 simple/redis'
```

Comment:
Split into two separate `after` blocks because it looks like rspec handles exceptions in `after` blocks nicely: it will still continue executing any remaining `after` blocks for cleanup. This means that the `stack rm --force simple` will still happen even if the `kontena service unlink` fails for some reason, allowing the remaining specs to run:

```
service link
  with a test-1 service
    returns error if target does not exist
    with a test-2 service
      links service to target
    with a simple stack
      links service to stack service (FAILED - 1)
    with a simple stack that has existing links to service
      links stack service with existing links to service

Failures:

  1) service link with a test-1 service with a simple stack links service to stack service
     Got 0 failures and 2 other errors:

     1.1) Failure/Error: raise Error.new(cmd, k.code, k.out) unless k.code.zero?
          
          Shell::Error:
            command failed with code 1: kontena service link test-1x simple/redis
             [error] 404 : Not found (/v1/services/e2e/null/test-1x)
          # ./spec/support/shell.rb:28:in `block in run!'
          # ./spec/support/shell.rb:27:in `tap'
          # ./spec/support/shell.rb:27:in `run!'
          # ./spec/features/service/link_spec.rb:43:in `block (4 levels) in <top (required)>'

     1.2) Failure/Error: raise Error.new(cmd, k.code, k.out) unless k.code.zero?
          
          Shell::Error:
            command failed with code 1: kontena service unlink test-1 simple/redis
             [error] Service is not linked to simple/redis
          # ./spec/support/shell.rb:28:in `block in run!'
          # ./spec/support/shell.rb:27:in `tap'
          # ./spec/support/shell.rb:27:in `run!'
          # ./spec/features/service/link_spec.rb:39:in `block (4 levels) in <top (required)>'

Top 4 slowest examples (14.92 seconds, 100.0% of total time):
  service link with a test-1 service with a test-2 service links service to target
    4.9 seconds ./spec/features/service/link_spec.rb:21
  service link with a test-1 service with a simple stack that has existing links to service links stack service with existing links to service
    4.2 seconds ./spec/features/service/link_spec.rb:63
  service link with a test-1 service with a simple stack links service to stack service
    3.45 seconds ./spec/features/service/link_spec.rb:42
  service link with a test-1 service returns error if target does not exist
    2.37 seconds ./spec/features/service/link_spec.rb:70

Finished in 14.92 seconds (files took 0.0979 seconds to load)
4 examples, 1 failure

Failed examples:

rspec ./spec/features/service/link_spec.rb:42 # service link with a test-1 service with a simple stack links service to stack service
```